Fix: only provide NAWS information from main console

SlySven opened the pull request:
I suspect that we might have been sending faulty information about `TConsole`s other than the main console.

This might close #5154.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight

Fixed a NAWS issue where Mudlet was telling the Game server about the sizes of windows other than the main one by mistake.

SlySven committed:
BugFix: only provide NAWS information from main console

I suspect that we might have been sending faulty information about
`TConsole`s other than the main console.

This might close #5154.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: try and cure odd NAWS messages with zero heights

This tries to fix #5154 by ensuring the values of `mScreenHeight` and
`mScreenWidth` from the upper pane of the `TMainConsole` are sent back to
the copies held in the `Host` class whenever they are modified as those are
the ones used in the NAWS messages.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

atari2600tim commented:
I can't tell if there's anything different here.  I have 2 windows open, one on the right is pr5160 and separate one on left is admin character watching the traffic from the 2 test characters.  Switching between them with a new size still sends a (40,0) whenever that profile gets selected.

https://user-images.githubusercontent.com/29287358/114798512-7da1a200-9d63-11eb-947c-42a7a4f08a14.mp4

keneanung reviewed (approved):
No user-visible text changed, so fine by me in ui-review manner.

vadi2 commented:
@SlySven could you look at https://github.com/Mudlet/Mudlet/pull/5160#issuecomment-819946898 ?

SlySven commented:
Ah, if a Window is hidden then the width and height would both be 0 however in `(void) TTextEdit::updateScreenView()` is the following fragment of code:
```cpp
    int currentScreenWidth = visibleRegion().boundingRect().width() / mFontWidth;
    if (mpConsole->getType() == TConsole::MainConsole) {
        // This is the MAIN console - we do not want it to ever disappear!
        mScreenWidth = qMax(40, currentScreenWidth);

        // Note the values in the "parent" Host instance - for the UPPER pane
        // so that they are available for NAWS:
        if (!mIsLowerPane) {
            mpHost->setScreenDimensions(mScreenWidth, mScreenHeight);
        }
    } else {
        mScreenWidth = currentScreenWidth;
    }
```
which forces the width of zero to be converted to `40` at a minimum - hence the `(40,0)` value. I think we need to:
either: not send any NAWS information when the height (of the upper pane) is zero
or: only send NAWS information when the main console is showing (i.e. it's profile tab is either the active one or multi-view is enabled)

I think the former will be simpler to implement.

SlySven committed:
BugFix: another detail to try and fix odd NAWS messages

This seems to be successful as it eliminates NAWS messages when the screen
height is zero - which seems to be when the particular profile's main
`TConsole` is hidden during non multi-view operation when more than one
profile is loaded. The odd NAWS messages about windows of size 40 x 0 are
now explained as there is some code to maintain a minimum width of 40 for
the main `TConsole` - which is taking effect even in the hidden case.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into BugFix_onlyProvideNAWSInformationFromMainConsole

Needed to bring in revisions to CI build processes.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

atari2600tim commented:
New build does no longer send the misleading 40,0 message when switching profiles, so improvement there.

But now it does not send anything when the profile is brought to the front.  So if you resize window with A in front and then switch to B, the game doesn't know B has a new size.

MultiView is great for both while it's open.  Resize window and A&B both send updates.  Exit multiview, A sends update but switch from A to B and game doesn't hear anything from B, still thinks it is half width.

Without looking through code, I'd speculate that whatever event happens at profile switching, where it had 0,0 and was applying a minimum and then sending 40,0, maybe still send something during that same event, but at a later step after it knows the number.  I mean the time when it sent 40,0, that wasn't the right number, but it was at a sensible time.  It was happening when it got brought to the front and was a different size than when it went to the back.

SlySven commented:
\**sigh*\*

Okay I will take another look...

SlySven committed:
Revise: check and potentially send NAWS data on all main console resizing

Now stores the main console dimensions and only send NAWS data if it
changes. This is partially done by overloading the resizeEvent() for that
main console.

Also rename some methods to better reflect their functionality and remove
buffer `TConsole` instances from some of the
`TConsole::resizeEvent(QResizeEvent*)` code.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: include check for Host pointer not being a nullptr

This keeps the GHA clang-tidy checker happy (makes unmodified code part in
a modified method pass the "clang-analyzer-core.NullDereference" check)!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: try again to keep GHA clang-tidy with checking for a valid pointer

This uses a slightly longer/more involved check than the prior effort
- hopefully it will be accepted when the last try just produced a more
lengthy warning message than before!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

atari2600tim commented:
Very much better now.  I opened two profiles like before and tried resizing with A in front and then switched to B and B sent an update as soon as it was selected.  I put it in multi-view mode and adjusted it which updated both.  Everything I tried last time is better now.  

It also remembers what it sent so that now you can drag it slowly without dragging far enough to change the number of characters and it won't spam the same number a bunch of times.  That's noticed and appreciated.

I did notice that when I turned on multi-view, only the left one sent an immediate update instead of both at the same time.  I clicked over in the command box on the right and sent a command and it still didn't send an update.  But then when I clicked on the tab at top then it sent an update.  I guess clicking the tab counts as switching profile but clicking and typing in it doesn't?  Control-Tab works great same as clicking the tab.

Otherwise all great except for the one weird case.

SlySven commented:
> ...  But then when I clicked on the tab at top then it sent an update. I guess clicking the tab counts as switching profile but clicking and typing in it doesn't?

:frowning:

Actually I just hooked into (in C++ terms I *overloaded*) the existing "Resize Event" for all `TConsole`s with a `TMainConsole` method just, as one might guess, for the main console. So i will see if something extra, just for the starting multi-view mode needs to be added...

mpconley commented:
This will be a very helpful change -Tamarindo

SlySven committed:
Merge branch 'development' into BugFix_onlyProvideNAWSInformationFromMainConsole

Needed so that GHA Windows CI builds will work.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into BugFix_onlyProvideNAWSInformationFromMainConsole

Needed to pull in development branch changes that include a Lua testing
framework.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

atari2600tim commented:
> I guess clicking the tab counts as switching profile but clicking and typing in it doesn't? Control-Tab works great same as clicking the tab.

Adding it to `mudlet::activateProfile` seems to get it where it will happen upon clicking in the input box.

mpconley reviewed (commented):
Does this PR need some infrastructure love to get it merge-able? I was not able to verify it with MacOS. Very hopeful this could make it for the next release, it is a good fix.

vadi2 committed:
Merge branch 'development' into BugFix_onlyProvideNAWSInformationFromMainConsole

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -743,42 +743,53 @@ bool cTelnet::socketOutRaw(std::string& data)
+    int naws_x = (pHost->mScreenWidth < pHost->mWrapAt) ? pHost->mScreenWidth : pHost->mWrapAt;
+    int naws_y = pHost->mScreenHeight;
+    if ((naws_y > 0) && (myOptionState[static_cast<size_t>(OPT_NAWS)]) && ((mNaws_x != naws_x) || (mNaws_y != naws_y))) {
+        qDebug().nospace().noquote() << "cTelnet::checkNAWS()() INFO - want to send NAWS data (for: \"" << mProfileName << "\") of X = " << naws_x << ", Y = " << naws_y << ".";
```
Let's remove this debug echo before merging.

SlySven replied:
```suggestion
```

vadi2 reviewed (approved):
I think it was just the cpp team, infrastructure was good. How come you weren't able to test this?

Approving just from the C++ side of things, I am not versed in NAWS so not able to test this.

demonnic commented:
fixed title and reran danger checks, which clear now

atari2600tim commented:
> > > ...  But then when I clicked on the tab at top then it sent an update. I guess clicking the tab counts as switching profile but clicking and typing in it doesn't?
> > 
> > 😦
> > 
> > Actually I just hooked into (in C++ terms I _overloaded_) the existing "Resize Event" for all `TConsole`s with a `TMainConsole` method just, as one might guess, for the main console. So i will see if something extra, just for the starting multi-view mode needs to be added...

> Adding it to `mudlet::activateProfile` seems to get it where it will happen upon clicking in the input box.

I should have mentioned that I did a pull request https://github.com/SlySven/Mudlet/pull/41 making that change.  With this change, it does send NAWS when you click into the other profile without using the tab in MultiView.

SlySven commented:
Just spotted this - had to approve the GHActions to run on it on my repository - just doing that and will merge that in and then take the debug line out...

atari2600tim committed:
put update dimensions in activateProfile (SlySven#41)

not sure if this is ideal spot, but this seems to do it

Authored by: Tim Johnson <cmdrkeen2@gmail.com>

SlySven committed:
Update src/ctelnet.cpp

Remove temporary debugging message.

SlySven commented on src/ctelnet.cpp:
```diff
@@ -743,42 +743,53 @@ bool cTelnet::socketOutRaw(std::string& data)
+    int naws_x = (pHost->mScreenWidth < pHost->mWrapAt) ? pHost->mScreenWidth : pHost->mWrapAt;
+    int naws_y = pHost->mScreenHeight;
+    if ((naws_y > 0) && (myOptionState[static_cast<size_t>(OPT_NAWS)]) && ((mNaws_x != naws_x) || (mNaws_y != naws_y))) {
+        qDebug().nospace().noquote() << "cTelnet::checkNAWS()() INFO - want to send NAWS data (for: \"" << mProfileName << "\") of X = " << naws_x << ", Y = " << naws_y << ".";
```
@vadi2 How about this one - it needs a typo cleaning up anyhow?
```suggestion
        qDebug().nospace().noquote() << "cTelnet::checkNAWS() INFO - want to send NAWS data (for: \"" << mProfileName << "\") of X = " << naws_x << ", Y = " << naws_y << ".";
```
